Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+in response. In some cases, this may be useful. Perhaps you want to send an email based on a content item and its template, but
+you must be very careful not to use this for purposes it is not meant to be used for.
+
+Generally speaking this method should not be used for the normal rendering of content. If abused this could cause severe
+performance problems. For normal content rendering of module type data from another content item, you should use Partial Views instead.
+
+## Don't put logic inside your constructors
+
+Constructors should generally not perform any logic, they should set some parameter values, perform some null checks and perhaps validate some data but in most cases, they should not perform any logic.
+
+There's a few reasons why this can become a huge performance problem:
+
+* The consumer of an API doesn't expect that by creating an object that they should be worried about performance
+* Creating an object can inadvertently happen a vast number of times, especially when using LINQ
+
+Here's an example of how this can go wrong very quickly:
+Your tree structure is something like this:
+
+    - Root
+    -- Home
+    --- Recipes (node id = 3251, list view with 5000 nodes)
+    --- About Us
+    --- Contact Us
+
+You have a custom model that looks like:
+
+```csharp
+public class RecipeModel : PublishedContentWrapped
+{
+    public RecipeModel(IPublishedContent content, IPublishedValueFallback publishedValueFallback) : base(content, publishedValueFallback)
+    {
+        RelatedRecipes = content
+            .Parent
+            .Children<RecipeModel>()
+            .Where(x => x.Value<IEnumerable<int>>("related")
+                .Contains(content.Id));
+
+        Votes = content.Value<int>("votes");
+    }
+    
+    public int Votes { get; private set; }
+    
+    public IEnumerable<RecipeModel> RelatedRecipes { get; private set; }
+}
+}
+```
+
+You then run the following code to show to show the favorites
+
+```csharp
+@var recipeNode = Umbraco.TypedContent(3251);
+@{
+	var recipeNode = Umbraco.Content(1234);
+}
+
+<ul>
+	@foreach (var recipe in recipeNode.Children
+		.Select(x => new RecipeModel(x, _publishedValueFallback))
+		.OrderByDescending(x => x.Votes)
+		.Take(10))
+	{
+		<li><a href="@recipe.Url()">@recipe.Name</a></li>
+	}	
+</ul>
+
+```
+
+__Ouch!__ To show the top 10 voted recipe's this will end up doing the following:
+
+* This will iterate over all Recipes, create and allocate 5000 instances of `IPublishedContent`
+* This will create and allocate 5000 instances of `RecipeModel`
+* For each `RecipeModel` created, this will traverse upwards, iterate all 5000 recipes then resolve property data for 2 properties
+
+This means that there is now a minimum of __20,000__ new objects created and allocated in memory. The number of traversals/visits to each
+of these objects is now 5000 x 5000 = __25,000,000 (25 MILLION TIMES!)__
```

Comment:
```suggestion
__Ouch!__ To show the top 10 voted recipes's this will end up doing the following:

* This will iterate over all Recipes, create and allocate 5000 instances of `IPublishedContent`
* This will create and allocate 5000 instances of `RecipeModel`
* For each `RecipeModel` created, this will traverse upwards, iterate all 5000 recipes then resolve property data for 2 properties

This means that there is now a minimum of __20,000__ new objects created and allocated in memory. The number of traversals/visits to each
of these objects is now 5000 x 5000 = __25,000,000 (25 MILLION TIMES!)__
```